File: src/testharness/signal-test-builder.ts
import type { Signal } from "../domain/value-objects/signal.js";

export class SignalTestBuilder {
  private payload = "default prompt";
  private source = "test";
  private sessionId: string | undefined;

  private constructor() {}

  static create(): SignalTestBuilder {
    return new SignalTestBuilder();
  }

  withPayload(payload: string): SignalTestBuilder {
    this.payload = payload;
    return this;
  }

  withSource(source: string): SignalTestBuilder {
    this.source = source;
    return this;
  }

  withSessionId(sessionId: string): SignalTestBuilder {
    this.sessionId = sessionId;
    return this;
  }

  build(): Signal {
    return {
      payload: this.payload,
      source: this.source,
      ...(this.sessionId !== undefined ? { sessionId: this.sessionId } : {}),
    };
  }
}

File: src/haas/HaaS.Domain.Tests/Builders/SignalTestBuilder.cs
using HaaS.Domain.ValueObjects;

namespace HaaS.Domain.Tests.Builders;

public class SignalTestBuilder
{
    private string _payload = "default prompt";
    private string _source = "test";
    private string? _sessionId;

    private SignalTestBuilder() { }

    public static SignalTestBuilder Create() => new();

    public SignalTestBuilder WithPayload(string payload)
    {
        _payload = payload;
        return this;
    }

    public SignalTestBuilder WithSource(string source)
    {
        _source = source;
        return this;
    }

    public SignalTestBuilder WithSessionId(string sessionId)
    {
        _sessionId = sessionId;
        return this;
    }

    public Signal Build() => new(_payload, _source, _sessionId);
}
